### PR TITLE
net:  Simplify the semantics around MTU

### PR DESCRIPTION
**Pull request by sboeuf**
And add the vhost-user-net support for the MTU parameter.

### CODE REVIEW

File: virtio-devices/src/net.rs [hunk 1]
```diff
@@ -565,6 +566,12 @@ impl Net {
             taps.push(tap);
         }
 
+        if let Some(mtu) = mtu {
+            if !taps.is_empty() {
+                taps[0].set_mtu(mtu as i32).map_err(Error::TapError)?;
```
Comment (rbradford) [inline virtio-devices/src/net.rs:571]: Only setting on the first one? What about the others?

File: virtio-devices/src/net.rs [hunk 2]
```diff
@@ -545,6 +545,7 @@ impl Net {
         id: String,
         fds: &[RawFd],
```
Comment (rbradford) [inline virtio-devices/src/net.rs:546]: nit: typo of default in commit message. I also think your commit message doesn't say what the code does or why

```
net: Don't override default TAP interface MTU

Adjust MTU logic such that:
1. Apply an MTU to the TAP interface if the user supplies it
2. Always query the TAP interface for the MTU and expose that.
```

File: virtio-devices/src/net.rs (hunk 1 above)
Comment (rbradford) [reply virtio-devices/src/net.rs:571]: Also, you have a code that checks if `taps` is empty and if so conditionally supply the feature. Can taps ever actually be empty? Should it not instead error out.

File: virtio-devices/src/net.rs (hunk 2 above)
Comment (sboeuf) [reply virtio-devices/src/net.rs:546]: Ok!

File: virtio-devices/src/net.rs (hunk 1 above)
Comment (sboeuf) [reply virtio-devices/src/net.rs:571]: Isn't the MTU at the interface level? I mean we don't set IP and MAC address on each queue, only on the first one.

Comment (sboeuf) [reply virtio-devices/src/net.rs:571]: > Also, you have a code that checks if `taps` is empty and if so conditionally supply the feature. Can taps ever actually be empty? Should it not instead error out.

Sure we could simplify the whole thing by returning an error if it's actually empty, and after that we can assume it's not empty.

Comment (rbradford) [reply virtio-devices/src/net.rs:571]: > Sure we could simplify the whole thing by returning an error if it's actually empty, and after that we can assume it's not empty.

This is what I would do - the vector shouldn't be empty.

File: vhost_user_net/src/lib.rs [hunk 3]
```diff
@@ -184,7 +188,13 @@ impl VhostUserBackendMut<VringRwLock<GuestMemoryAtomic<GuestMemoryMmap>>, Atomic
             | 1 << VIRTIO_NET_F_MAC
             | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
             | 1 << VIRTIO_F_VERSION_1
-            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
+
+        if self.mtu.is_some() {
```
Comment (rbradford) [inline vhost_user_net/src/lib.rs:193]: This feature should always be advertised - we can always get an MTU from the tap.

Comment (sboeuf) [reply vhost_user_net/src/lib.rs:193]: Yep!

File: virtio-devices/src/net.rs [hunk 4]
```diff
@@ -162,6 +162,8 @@ pub enum Error {
     TapError(TapError),
     #[error("Error calling dup() on tap fd: {0}")]
     DuplicateTapFd(std::io::Error),
+    #[error("No tap device, at least one should be created")]
```
Comment (rbradford) [inline virtio-devices/src/net.rs:165]: This isn't about taps provided by the user. This is a tap device missing when creating the network.  If there is no reason a user can trigger it then it should be an assert! i.e. programming error.

File: vmm/src/config.rs [hunk 5]
```diff
@@ -1247,8 +1247,8 @@ pub struct NetConfig {
     pub mac: MacAddr,
     #[serde(default)]
     pub host_mac: Option<MacAddr>,
-    #[serde(default = "default_netconfig_mtu")]
```
Comment (likebreath) [inline vmm/src/config.rs:1250]: Don't forget to remove the default value from the openapi yaml file.

File: virtio-devices/src/net.rs (hunk 4 above)
Comment (sboeuf) [reply virtio-devices/src/net.rs:165]: Sounds good, I'll use an `assert!()`.

File: vmm/src/config.rs (hunk 5 above)
Comment (sboeuf) [reply vmm/src/config.rs:1250]: Thanks!

**Pull request merged**